### PR TITLE
Question Controller and views for Expanded Q & A system

### PR DESCRIPTION
**Pull request by ananyo2012**
This contains the first two tasks listed in Issue #549 

* [x] Write basic CRUD actions for Question Controller and corresponding Views
* [x] Write functional tests for Question Controller
* [x] tests pass -- `rake test`
* [x] code is in uniquely-named feature branch, and has been rebased on top of latest master (especially if you've been asked to make additional changes)
* [x] pull request are descriptively named
* [x] if possible, multiple commits squashed if they're smaller changes
* [ ] reviewed/confirmed/tested by another contributor or maintainer
* [x] `development.sqlite.example` has been updated if any database migrations were added

### CODE REVIEW

**Comment by jywarren** [comment]
Hi, Ananyo, thanks -- i'm guessing this is not ready for review? If it's not, can you uncheck some boxes, since I look at the # of steps complete (in list view) to tell if I should try merging? Thanks! And/or will you simply rename it and update when it becomes ready?

Thanks.

**Comment by ananyo2012** [comment]
Here are the first looks of the Q & A Pages. Picture below shows the view of Questions index Page.
question_index

This is the view of the question show page
question_show

**Comment by ananyo2012** [comment]
Before this could be merged i need to solve one issue which is, while the user is is not logged in Clicking the Ask a Question button redirects him to the login page but the redirection url after logging in doesn't contain the question parameters in the url.

**Comment by ananyo2012** [comment]
@jywarren @liz Please have a look at the initial changes made.

**Comment by ananyo2012** [comment]
Since question are actually a modified research notes so actions involving post requests are same as that of research notes just their redirection url will differ so I have modified notes controller so that is receives a `:redirect` parameter and redirects to the required url accordingly.

**Comment by jywarren** [comment]
Thanks, do you think we could use the new question templates from the
dashboard redesign?

And we should have an area at the top of this page where it describes the
page and prompts you to ask a question using a button or form. Thanks!

**Comment by ananyo2012** [comment]
@jywarren Yeah perhaps we could. They look good. Should I use them? I have already added a button for posting question. I think showing tags in the index page would be good too.

**Comment by jywarren** [comment]
I guess I'd like to think about some different page designs, and I'll do
some sketching for Tuesday. But yes, in the meantime, do you think you
could rework your code to use the newer templates?

Thanks!

**Comment by ananyo2012** [comment]
Ok Jeff I will use the templates and modify the code.

**Comment by ananyo2012** [comment]
@jywarren Please review the changes made. It's almost ready for merge. Please suggest changes needed. I have used the question template used in dashboard page as suggested.

**Comment by ananyo2012** [comment]
This is how the page listing all questions looks now.
questions_index_v2

**Comment by jywarren** [comment]
@ebarry @steviepubliclab -- I think we can merge this as it will be visible to all, but not prominently linked-to from anywhere and we can continue to refine the design as we go. It'll be good to get some of this great work into the codebase, as we're encouraging students to merge in their new changes regularly instead of saving them all up.

**Comment by jywarren** [comment]
I'm merging this now, and will post some additional sketches hopefully later tonight or tomorrow for next steps on design. Ananyo -- looks great and I think the design side can progress in parallel to some of the coding work you're continuing. Thanks!

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -80,7 +80,7 @@ def require_user
       unless current_user
         store_location
         flash[:notice] = "You must be logged in to access this page"
-        redirect_to login_url+'?return_to=' + URI.encode(request.env['PATH_INFO'])
```
Comment (jywarren) [inline app/controllers/application_controller.rb]: Actually, wait - why was this removed?

**Comment by jywarren** [comment]
Hi, Ananyo -- I'm sorry, I just reverted this and wanted to ask for some changes. I'm very sorry to have to have entered the reversion in the history, and apologize -- I should've been a bit more careful. The code is close to ready but I want to ask for just a few more changes first. I'll leave inline questions.

File: app/controllers/notes_controller.rb [hunk 2]
```diff
@@ -103,7 +103,11 @@ def create
         else
           flash[:notice] = "Research note published. Get the word out on <a href='/lists'>the discussion lists</a>!"
         end
-        redirect_to @node.path
+        if params[:redirect] && params[:redirect] == 'question'
```
Comment (jywarren) [inline app/controllers/notes_controller.rb]: Instead of doing this here, can we do it inside the `@node.path` method, which we already use to switch between types? I think we should be able to based on whether there are tags on the note, since they'll have been added by this point.

File: app/controllers/questions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    @title = @node.latest.title
+    @tags = @node.tags
+    @tagnames = @tags.collect(&:name)
+
+    set_sidebar :tags, @tagnames
+  end
+
+  def shortlink
+    @node = DrupalNode.find params[:id]
+    redirect_to '/questions/'+@node.author.name+'/'+Time.at(@node.created).strftime("%m-%d-%Y")+'/'+@node.title.parameterize
```
Comment (jywarren) [inline app/controllers/questions_controller.rb]: Is it possible to use `@node.path` here too?

Comment (jywarren) [inline app/controllers/questions_controller.rb]: This helps centralize where paths are generated.

File: app/views/editor/post.html.erb [hunk 4]
```diff
@@ -79,6 +79,10 @@
         <h3>Share your work</h3>
       <% end %>
   
+      <% if params[:redirect] == 'question' %>
+        <%= hidden_field_tag 'redirect', 'question' %>
```
Comment (jywarren) [inline app/views/editor/post.html.erb]: Can we instead simply rely on the provided tagging?

File: app/views/like/_like.html.erb [hunk 5]
```diff
@@ -33,7 +33,7 @@
   <hr />
 
   <div class='btn-group'>
-    <% if node.type == 'note' %><a <% if current_user && current_user.uid != node.uid %>data-confirm='Please be aware that you are editing someone else\'s post, a privilege you have only because you are a moderator or admin.' <% end %>class='btn btn-default btn-sm' href='/notes/edit/<%= node.id %>?t=<%= Time.now.to_i %>'><i class='fa fa-pencil'></i><span class='hidden-xs hidden-sm'> Edit</span></a><% end %>
+    <% if node.type == 'note' %><a <% if current_user && current_user.uid != node.uid %>data-confirm='Please be aware that you are editing someone else\'s post, a privilege you have only because you are a moderator or admin.' <% end %>class='btn btn-default btn-sm' href='/notes/edit/<%= node.id %>?t=<%= Time.now.to_i %><% if params[:controller] == "questions" %>&redirect=question<% end %>'><i class='fa fa-pencil'></i><span class='hidden-xs hidden-sm'> Edit</span></a><% end %>
```
Comment (jywarren) [inline app/views/like/_like.html.erb]: We should do this at the controller level, as it's relatively hidden here. Can we check for tagging at the controller and redirect there?

File: app/views/sidebar/_post_button.html.erb [hunk 6]
```diff
@@ -1,10 +1,22 @@
+<% if params[:controller] == 'questions' %>
```
Comment (jywarren) [inline app/views/sidebar/_post_button.html.erb]: I'd like to keep this template file simpler. Can we put this question button at the top of the questions index page?

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (ananyo2012) [inline app/controllers/application_controller.rb]: Actually the `store_location` method already stores the full path of the request in the session. See reference here: https://github.com/publiclab/plots2/pull/550/files#diff-55c5b7aecfb519d0e4880eaf2788eb6eL98. The `URI.encode(request.env['PATH_INFO'])` was actually returning the parameters but it was only available to `\login` page. As soon it posted the login parameters of an user the parameters were no longer available and it didn't redirect to with no parameters in the `/post` url. So it was an issue that I solved it here using the changes done in the user_sessions_controller: https://github.com/publiclab/plots2/pull/550/files#diff-0770692f348632e109afe78e3585bfe0R35

File: app/controllers/notes_controller.rb (hunk 2 above)
Comment (ananyo2012) [inline app/controllers/notes_controller.rb]: Oh yes we can surely do that by changing the path for questions in the `generate_path` method. But that would mean users could no longer view the as normal research notes. Thus it would affect the present case if deployed on the main website, thus affecting the backward compatibility. So I devised a way so that question could co-exist in the research notes page as well as in the questions page.

File: app/controllers/questions_controller.rb (hunk 3 above)
Comment (ananyo2012) [inline app/controllers/questions_controller.rb]: Similar answer as the previous thread.

File: app/views/editor/post.html.erb (hunk 4 above)
Comment (ananyo2012) [inline app/views/editor/post.html.erb]: This is added to provide redirect to the question show path only when a `:redirect` parameter is given in the url. Well I thought of redirecting the questions to the notes show page if questions are posted from post a research note link so that the user doesn't gets redirected to the questions show page which is currently under construction.

File: app/views/like/_like.html.erb (hunk 5 above)
Comment (ananyo2012) [inline app/views/like/_like.html.erb]: Similar answer as the last thread.

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (jywarren) [inline app/controllers/application_controller.rb]: I see -- great, thanks!

Comment (jywarren) [inline app/controllers/application_controller.rb]: In the future, perhaps smaller, independent changes could be submitted in separate pull requests, however, where possible. Thanks, though, this is a good catch and a welcome contribution!

File: app/views/sidebar/_post_button.html.erb (hunk 6 above)
Comment (ananyo2012) [inline app/views/sidebar/_post_button.html.erb]: Ok Jeff I will add the **Ask a question** button at the top of the question show page and index page.

File: app/controllers/notes_controller.rb (hunk 2 above)
Comment (jywarren) [inline app/controllers/notes_controller.rb]: I see! Great to know this was considered. I'm interested in how this can be clearly communicated in how the code is implemented, and also keep path generation code centralized.

Do you think we could have `@node.path()` accept an optional parameter, like `def path(type = :default)` (best double check that for setting defaults as `:symbols` but I think it'd work) where using `@node.path(:question)` would generate the question version of the path? That also reduced redundancy in this controller.

File: app/views/editor/post.html.erb (hunk 4 above)
Comment (jywarren) [inline app/views/editor/post.html.erb]: I see -- so this is temporary? If so, let's leave a clear comment that it should be removed once question show template is complete. Something like `<!-- Notice: Remove this ... -->`. Thanks!

File: app/controllers/application_controller.rb (hunk 1 above)
Comment (ananyo2012) [inline app/controllers/application_controller.rb]: Yeah this is a issue I should have handled separately raising a github issue.Will do it as per from the next time. Thanks Jeff!

File: app/views/editor/post.html.erb (hunk 4 above)
Comment (ananyo2012) [inline app/views/editor/post.html.erb]: Ok will add comments wherever necessary.
